[Plug N Play] All the pieces reconnected together - Top level API functioning ready for integration in PIP

DEKHTIARJonathan opened the pull request:
- Align the names [there was a bunch of name with different forms]:
    - namespace
      - namespace_priority
      - namespace_priorities **=> Aligned to this**
      - namespaces_priority
    - feature
      - feature_priority
      - feature_priorities **=> Aligned to this**
      - features_priority
    - property
      - property_priority
      - property_priorities **=> Aligned to this**
      - properties_priority
      
- The idea that `get_combinations` can output a sorted order is:
1. Flawed in nearly all cases because it's dynamic depending on user inputs/configurations
2. Possible with extreme complexity
3. Completely unnecessary it's a testing it needs to stay simple.

- Moved the `unpack_variants_json` in a separate file (AFAIK not a public API)

- `get_variant_hashes_by_priority` is now feature-complete AFAIK (supports user-supplied priorities / forbidden objects).

- `VariantDescription` objects can now do: `from_dict()` and `to_dict()` => useful to go IN/OUT of `variants.json`
    - used both in `api.py` and `generate_index_json`

mgorny commented on tests/test_api.py:
```diff
@@ -187,19 +130,33 @@ def test_get_variant_hashes_by_priority_roundtrip(configs):
         for i, x in enumerate(get_combinations(configs)):
-            assume(i < 65536)
+            if i >= 65536:
+                pytest.skip("Too many combinations")
```
Why are you changing this? `assume()` helps hypothesis choose better samples, `skip()` just kills the test entirely.

DEKHTIARJonathan replied:
It's apparently deprecated. I got a warning message. Replace it back with assume you'll see the message

mgorny replied:
No, it's not. I'm guessing you got a warning because you added it to `test_get_variant_hashes_by_priority_roundtrip()` which doesn't use Hypothesis, and it makes sense only when you are actually using Hypothesis.

DEKHTIARJonathan replied:
Ahhh that would explain. My bad I'll revert :)

mgorny commented on tests/test_api.py:
```diff
@@ -44,92 +45,34 @@ def configs(mocked_plugin_loader: type[PluginLoader]):  # noqa: F811
-    assert isinstance(provider_cfgs, (list, tuple))
-    assert len(provider_cfgs) > 0
-    assert all(isinstance(config, ProviderConfig) for config in provider_cfgs)
+    def get_or_skip_combinations() -> Generator[VariantDescription]:
```
Why do you need that? The test data is static, so the combination count is static too, so you're either never or always skipping the test.

DEKHTIARJonathan replied:
True. Removing

mgorny commented on variantlib/commands/generate_index_json.py:
```diff
@@ -91,34 +99,45 @@ def generate_index_json(args: list[str]) -> None:
+                },
             },
             f,
+            indent=4,
```
Well, the initial idea is that `variants.json` was supposed to be as small as possible, and this goes in the opposite direction.

DEKHTIARJonathan replied:
OIC. Though does it really matter to save 1kb and make it more readable ? How about an option `--compact` or `--beautify` (depending which is default).

It's quite hard to read/debug when it's all in one line :/

mgorny replied:
I don't mind changing it. Just pointing out that the feeling I got from the initial discussion is that we wanted to add as little space as possible — but perhaps that was because you didn't like the file back then. I've been using `json_pp` to format it when I needed to read it.

mgorny commented on variantlib/models/variant.py:
```diff
@@ -111,12 +112,12 @@ class VariantProperty(VariantFeature):
         return hash((self.__class__, self.namespace, self.feature, self.value))
 
-    @cached_property
+    @property
```
I don't think this is a good use of property. The returned object is not a property (i.e. something belonging to) of the original `VariantProperty` but rather a completely new, independent object. The usual way of constructing new objects in Python is via method and not property.

Also, it would be more consistent to follow the naming scheme of `to_*` since it's converting another type.

DEKHTIARJonathan replied:
What about something like `pathlib.Path().parent` ? It does also return a new `Path()`

Like `.parent_feature` - I don't like much `feature_object` taking ideas if you have smthg in mind.

If you really want it as a function we can - this API is pretty much internal-only.

mgorny replied:
Yeah, that's a fair point. The stdlib is already inconsistent, so I guess it's fine either way. (but then, `to_str()` could be a property too, and that would be mightily confusing.)

mgorny commented on variantlib/utils.py:
```diff
@@ -1,8 +1,30 @@
+T = TypeVar("T")
+
+
+def aggregate_user_and_default_lists(
```
This can either mutate `user_list` in place or return mutable `default_list`. You should either copy the lists, or at least note the mutability in the docstring, since it can be confusing, especially if `default_list` ends up being a global/shared list somewhere.

DEKHTIARJonathan replied:
Oh good point. I don't like mutability. Makes bug tracking very hard to track

DEKHTIARJonathan replied:
Fixed

mgorny commented on variantlib/variant_file.py:
```diff
@@ -0,0 +1,28 @@
+
+
+def unpack_variants_json(
+    variants_json: dict[str, dict[str, dict]],
```
```suggestion
    variants_json: dict,
```

`variants.json` can have other top-level keys whose values may not be `dict`s.

DEKHTIARJonathan replied:
Ah right

mgorny reviewed (changes requested):
Furthermore, I disagree with killing sorted `get_combinations()`. Sure, it does not handle all the possible use cases, **because it was never supposed to**.

It is a second implementation of the ordering algorithm for very specific circumstances (namespaces ordered by priority, no feature/value priorities, no filtering). It is valuable as an independent way of fuzzing the sort algorithm, because it is completely independent of the main sort algorithm. If we start making changes into the algorithm, these changes may not happen to surface in fixed test cases, and that is precisely the purpose of fuzzing.

DEKHTIARJonathan commented:
@mgorny I added back the original implementation of `get_combinations` with a little twist (I added `namespace_priority`) to allow some fun with it a little bit. Added some more unittests since I was there.

mgorny commented on tests/test_api.py:
```diff
@@ -22,9 +22,11 @@
 from variantlib.loader import PluginLoader
 from variantlib.models import provider as pconfig
 from variantlib.models import variant as vconfig
+from variantlib.models.configuration import VariantConfiguration as VConfigurationModel
```
I don't really understand why the rename.

mgorny commented on tests/test_api.py:
```diff
@@ -44,92 +46,30 @@ def configs(mocked_plugin_loader: type[PluginLoader]):  # noqa: F811
+        "variantlib.configuration.VariantConfiguration.get_config"
+    ).return_value = VConfigurationModel(namespace_priorities=namespace_priorities)
 
+    result = set(get_variant_hashes_by_priority(variants_json=variants_json))
+    combination_vhashs = {vdesc.hexdigest for vdesc in combinations}
+    assert len(result) == len(combination_vhashs)
```
Can't we check the ordering here, given we only do namespace priorities?

mgorny replied:
Yep, readded that, and it works.

mgorny commented on tests/test_api.py:
```diff
@@ -187,19 +127,32 @@ def test_get_variant_hashes_by_priority_roundtrip(configs):
-        x.hexdigest for x in combinations
-    ]
+
+    result = set(get_variant_hashes_by_priority(variants_json=variants_json))
+    combination_vhashs = {vdesc.hexdigest for vdesc in combinations}
+    assert len(result) == len(combination_vhashs)
+
+    assert result.symmetric_difference(combination_vhashs) == set()
```
Same here.

mgorny commented on tests/test_utils.py:
```diff
@@ -0,0 +1,151 @@
+
+import pytest
+
+from tests.test_plugins import mocked_plugin_loader  # noqa: F401
```
At this point, I think we should move it to pytest. But that's to be done separately, I'll add to #23.

mgorny commented on tests/test_utils.py:
```diff
@@ -0,0 +1,151 @@
+
+
+@pytest.fixture
+def configs(mocked_plugin_loader: type[PluginLoader]):  # noqa: F811
```
```suggestion
def configs(mocked_plugin_loader: type[PluginLoader]) -> list[ProviderConfig]:  # noqa: F811
```
